### ISSUE TITLE
RPC API: calculation of available account balance is difficult or impossible

### ISSUE THREAD

**Issue by buck54321** [zcash/zcash#6679]
I've encountered what I think are inconsistencies around account handling that make the RPC API difficult to use for account balance calculations.

#### My use case
We're doing atomic swaps with Zcash. By the nature of atomic swaps, the addresses in the redemption script must be transparent so that you can audit them on-chain. At a base level, we have no use for shielded addresses/receivers, but I've been expanding the wallet features exposed through our multi-wallet UI so that users can work with shielded pools too (just for general use, not for swapping). For Zcash, we require a full node and we connect over RPC.

When zcashd deprecated `getnewaddress`, I migrated to accounts and to using `z_getaddressforaccount`, which returns a unified address (that **requires** a shielded receiver that we'll never use for swaps, but that's fine). We pull out the transparent receiver to communicate to the server and counterparty. For privacy, we generate new addresses for every swap.

#### The issue
For simplicity, I'm assuming this is not a mining wallet, but I make no assumptions about whether the user has balance in the legacy account. 

My hangup is with calculating **available** account transparent balances, and please forgive me if I'm missing something obvious about the API. Let's say the user wants to send some amount from their transparent funds to Orchard. I want to be able to tell the user what they have available to send. My naive approach was to use `getbalance`, subtract funds locked in `listlockunspent`, and present the result as the balance available for transfer. I would then use `z_sendmany` with the special string `ANY_TADDR` as the `fromaddress` field to send the funds to an orchard address. But [I now understand](https://github.com/zcash/zcash/issues/6654#issuecomment-1544773223) that `ANY_TADDR` does not pull from transparent outputs across all accounts, only the legacy account. Even though `getbalance`, `listunspent`, `listlockunspent` all consider transparent outputs across all accounts, `z_sendmany` has no way to pull from all accounts. I thought I might work around it by getting the balance available in each account separately and either only exposing a certain account's balance or sending two transactions if needed. I might also offer the option for the user to "upgrade" by moving funds from legacy to a numbered account. But how much is in the legacy account? There's no easy way to tell. I could jump through some hoops with `getbalance` and then calling `z_getbalanceforaccount` for every account and doing some math. But then I need subtract locked outputs, and neither `listunspent` nor `listlockunspent` accept an account number as a filter, nor do the results include which account number the outputs are from. I thought maybe I could check each output's address to see which account it comes from, but while `validateaddress` has an `account` field, it's a string field and it's deprecated. Nothing in `z_validateaddress` to help either.

#### tldr;
I can't find a way to tell calculate the **available** (not locked) balance for a single account, and no way to even query the balance in the "legacy" account.

#### Potential solutions
For myself, the cleanest solution would just be to allow `z_sendmany` to draw from the same pool of utxos considered in `getbalance` and `listunspent`. This would allow me to effectively treat legacy and account 0 as a single "transparent account". Maybe a new special argument `ANY_TADDR_ANY_ACCOUNT` could trigger this output selection mode. A better solution might be to add account numbers to relevant input args and account number fields to results. Maybe could use `-1` for the legacy account. But modifying the API like that would require me to write a fair amount of custom code and I'm lazy, so my vote is on a new selection option for `z_sendmany`.

**Comment by str4d** [comment]
> Even though `getbalance` [..] consider[s] transparent outputs across all accounts,

This is actually a bug that we introduced in version 4.7.0. We intended for the newly-introduced accounts to _not_ be considered part of the legacy wallet or APIs (that is, we wanted to _preserve_ the definition of "legacy transparent wallet" being a single bucket of funds, the way it has always been), but the eventual implementation didn't correctly restrict this. Fixing this is tracked by #5941; it's been near the top of our priority list since, but unfortunately just below all the work we've been doing to cope with the persistent network loads since June last year.

> My naive approach was to use `getbalance`, subtract funds locked in `listlockunspent`, and present the result as the balance available for transfer.

This was definitely never intended to work for accounts, and is unlikely to ever do so.

> I can't find a way to calculate the available (not locked) balance for a single account, and no way to even query the balance in the "legacy" account.

#6356 is the issue for this. We have a `z_getbalances` RPC draft in #6389 that you can try out and give feedback on. I don't believe it currently takes into account the locked status of notes, but that is an excellent use case that we should factor into the implementation.
